daml-script/dump/src/main/scala/com/daml/script/dump/Main.scala

```diff
@@ -57,29 +60,68 @@ object Main {
     ()
   }
 
+  private def getACS(
+      client: LedgerClient,
+      parties: List[String],
+      offset: LedgerOffset,
+  )(implicit
+      mat: Materializer
+  ): Future[Map[String, CreatedEvent]] = {
+    val ledgerBegin = LedgerOffset(
+      LedgerOffset.Value.Boundary(LedgerOffset.LedgerBoundary.LEDGER_BEGIN)
+    )
+    if (offset == ledgerBegin) {
+      Future.successful(Map.empty)
+    } else {
+      client.transactionClient
+        .getTransactions(ledgerBegin, Some(offset), filter(parties), verbose = true)
+        .runFold(Map.empty[String, CreatedEvent]) { case (acs, tx) =>
+          tx.events.foldLeft(acs) { case (acs, ev) =>
+            ev.event match {
+              case Event.Empty => acs
+              case Event.Created(value) => acs + (value.contractId -> value)
+              case Event.Archived(value) => acs - value.contractId
+            }
+          }
+        }
+    }
+  }
+
+  private def getTransactionTrees(
+      client: LedgerClient,
+      parties: List[String],
+      start: LedgerOffset,
+      end: LedgerOffset,
+  )(implicit
+      mat: Materializer
+  ): Future[Seq[TransactionTree]] = {
+    client.transactionClient
+      .getTransactionTrees(start, Some(end), filter(parties), verbose = true)
+      .runWith(Sink.seq)
+  }
+
   def run(config: Config)(implicit
       ec: ExecutionContext,
       esf: ExecutionSequencerFactory,
       mat: Materializer,
   ): Future[Unit] =
     for {
       client <- LedgerClient.singleHost(config.ledgerHost, config.ledgerPort, clientConfig)
-      trees <- client.transactionClient
-        .getTransactionTrees(
-          LedgerOffset(LedgerOffset.Value.Boundary(LedgerOffset.LedgerBoundary.LEDGER_BEGIN)),
-          Some(LedgerOffset(LedgerOffset.Value.Boundary(LedgerOffset.LedgerBoundary.LEDGER_END))),
-          filter(config.parties),
-          verbose = true,
-        )
-        .runWith(Sink.seq)
-      pkgRefs: Set[PackageId] = trees.toList
+      acs <- getACS(client, config.parties, config.start)
+      trees <- getTransactionTrees(client, config.parties, config.start, config.end)
+      acsPkgRefs = acs.values.toList
+        .foldMap(ev => valueRefs(Sum.Record(ev.getCreateArguments)))
+        .map(i => PackageId.assertFromString(i.packageId))
+      treePkgRefs = trees.toList
         .foldMap(treeRefs(_))
         .map(i => PackageId.assertFromString(i.packageId))
+      pkgRefs = acsPkgRefs ++ treePkgRefs
```

Comment:
Any reason not to work with sets here?

Reply:
Not sure I follow, they are `Set[PackageId]`.